src/gui/clipboard.cpp

```diff
@@ -0,0 +1,220 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "clipboard.h"
+
+#include "checks.h"
+#include "control.h"
+#include "logging.h"
+#include "string_utils.h"
+#include "unicode.h"
+
+#include <SDL.h>
+
+CHECK_NARROWING();
+
+static struct Clipboard {
+	// Content - text
+	uint16_t text_code_page = 0;
+	// Since DOS -> UTF-8 -> DOS string conversion won't necessary produce
+	// the same output as original DOS encoded text (some code pages has
+	// multiple blank/unused characters, etc.), in case the text comes from
+	// the DOS side we keep both DOS and UTF-8 encoded strings - this way
+	// the DOS can get precisely the same content as it has sent to the
+	// clipboard as long as the code page did not change in the meantime.
+	std::string text_utf8 = {}; // always filled in if we contain a text
+	std::string text_dos  = {}; // only set when pasting text from DOS
+
+	// Host content mirror
+	std::string host_text_utf8 = {};
+
+	void clear_content()
+	{
+		text_utf8.clear();
+		text_dos.clear();
+		text_code_page = 0;
+		// Do not clear 'host_text_utf8' - it is crucial to keep it
+		// intact for proper synchronization with host OS clipboard
+	}
+
+} clipboard;
+
+// ***************************************************************************
+// Host clipboard synchronization
+// ***************************************************************************
+
+static bool text_equals(const std::string& str_1, const std::string& str_2)
+{
+	// String compare, insensitive to end of the line encoding; needed
+	// because we convert the end of line endings to host native before
+	// copying the text to the host clipboard
+
+	size_t index_1 = 0;
+	size_t index_2 = 0;
+
+	auto get_character = [](const std::string& str, size_t& index) {
+		const char result = (str[index] == '\r') ? '\n' : str[index];
+		if ((index + 1 < str.size()) &&
+		    ((str[index] == '\r' && str[index + 1] == '\n') ||
+		     (str[index] == '\n' && str[index + 1] == '\r'))) {
+			// End of the line encoded as '\r\n' or '\n\r'
+			++index;
+		}
+
+		++index;
+		return result;
+	};
+
+	while ((index_1 < str_1.size()) && (index_2 < str_2.size())) {
+		if (get_character(str_1, index_1) != get_character(str_2, index_2)) {
+			return false;
+		}
+	}
+
+	return (index_1 == str_1.size()) && (index_2 == str_2.size());
+}
```

Comment:
We should have unit tests for this function.

Reply:
Done, unit test created.